include/libdnf/utils/exception.hpp

```diff
@@ -0,0 +1,76 @@
+/*
+Copyright (C) 2020 Red Hat, Inc.
+
+This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#ifndef LIBDNF_UTILS_EXCEPTION_HPP
+#define LIBDNF_UTILS_EXCEPTION_HPP
+
+
+#include <stdexcept>
+#include <system_error>
+
+namespace libdnf {
+
+/// Base class of libdnf exceptions. Each exception define at least domain_name, name, and description.
+class Exception : public std::runtime_error {
+public:
+    using runtime_error::runtime_error;
+    virtual const char * get_domain_name() const noexcept { return "libdnf"; }
+    virtual const char * get_name() const noexcept { return "Exception"; }
+    virtual const char * get_description() const noexcept { return ""; }
+};
+
+class LogicError : public Exception {
+public:
+    using Exception::Exception;
+    const char * get_name() const noexcept override { return "LogicError"; }
+    const char * get_description() const noexcept override { return "General LogicError exception"; }
+};
+
+class RuntimeError : public Exception {
+public:
+    using Exception::Exception;
+    const char * get_name() const noexcept override { return "RuntimeError"; }
+    const char * get_description() const noexcept override { return "General RuntimeError exception"; }
+};
+
+class SystemError : public RuntimeError {
+public:
+    SystemError(int code, const std::string & what) : RuntimeError(what), error_code(code) {}
+    SystemError(int code, const char * what) : RuntimeError(what), error_code(code) {}
+    const char * get_domain_name() const noexcept override { return "system"; }
+    const char * get_name() const noexcept override { return "SystemError"; }
+    const char * get_description() const noexcept override { return "System error"; }
+    const char * what() const noexcept override {
```

Comment:
Am I correct that you cannot get the original message that was used for constructor but only generated message from error code?

Reply:
@j-mracek Yes. You are correct. This is not first version of System error class. I done change in it and during change I done this bug. I will correct it.

Reply:
What about to add an additional method that will return only unformatted original message from what in constructor?
Or keep what() method for unformatted output but add additional method for formatted output? The rest LGTM.

Reply:
I made a change. The `what()` method returns formatted message - combined from  exception class description and messages in one string.
User can access the original message passed to the constructor by new method `get_message()`.

Reply:
I don't like having these multiple confusing methods on the interface. Between `get_description()`, `get_message()` and `what()`, without carefully reading the documentation (which we don't have yet but it needs to be good here), the user will be confused and will have no idea what is what (except, hopefully, for `what()` :sunglasses: but the additional methods will make them confused nonetheless). The names are mostly arbitrary and don't really express what they represent (and I don't think we can really do anything about the naming due to the whole nature of the concept).

Reply:
@lukash Please could you propose better function names?

Reply:
@j-mracek not really :unamused: there's no definition to them, it's just "one part of the error message" and "another part of the error message".

I, however, don't find it useful to expose these on the interface (I would just remove them and keep only `what()`). What use is part of the error message? We've discussed it with @jrohel already and he wants to keep at least the `get_description()`. That's not stopping me from pointing out the shortcomings though :slightly_smiling_face:

Reply:
@lukash 
> I, however, don't find it useful to expose these on the interface
So because you didn't find it useful, I have to hide it? If you don't see usage for these functions nobody can't use it? I have usage.
Example: Possibility to better translate (localize) message after catching exception. We discussed it. And even if we will not use this methods, somebody can use they. 

You said the user will be confused because there will be more methods available. But I made change and `what()` will be working as expected. Other methods are plus for users who want it. I met a poorly designed libraries. I wanted to get some result and only option was one long formatted string. And I had to parse it to found specific information. You don't have to use these function. But why are you fighting to ban others from using they?